etckeeper.8
.\" -*- nroff -*-
.TH ETCKEEPER 8 "" "" ""
.SH NAME
etckeeper \- store /etc in git, mercurial, bazaar, or darcs
.SH SYNOPSIS
.B etckeeper command [-d directory]
.SH DESCRIPTION
etckeeper manages /etc be stored in a git, mercurial, bazaar, or darcs
repository. By default each of the commands operates on /etc, but a
different directory can be specified to operate on a clone of the /etc
repository located elsewhere.
.SH COMMANDS
.TP
.B init
This initialises and sets up a git, mercurial, bazaar, or darcs
repository (depending on the VCS setting in
/etc/etckeeper/etckeeper.conf). Typically this is run in /etc once
when starting to use etckeeper on a machine. It can also be used to
initialise a clone of the /etc repository located elsewhere.
.TP
.B commit [message]
Commits all changes in /etc to the repository. A commit message can be
specified. You may also use the underlying VCS to commit manually.
(Note that etckeeper commit will notice if a user has used sudo or su to
become root, and record the original username in the commit.)
.TP
.B pre-commit
This is called as a pre-commit hook. It stores metadata and does sanity
checks.
.TP
.B pre-install
This is called by apt's DPkg::Pre-Install-Pkgs hook, or by equivalent hooks
of other package managers. It allows committing any uncommitted changes before
packages are installed, upgraded, etc.
.TP
.B post-install
This is called by apt's DPkg::Post-Invoke hook, or by equivalent hooks
of other package managers. It commits changes made by packages into the
repository. (You can also call this by hand after running dpkg by hand.)
.TP
.B unclean
This returns true if the directory contains uncommitted changes.
.TP
.B update-ignore [-a]
This updates the VCS ignore file. Content outside a "managed by etckeeper"
block is not touched. This is generally run when upgrading to a new version
of etckeeper. (The -a switch will add a "managed by etckeeper" block if
one is not present.)
.TP
.B vcs subcommand [options ...]
You can use this to run any subcommand of the VCS that etckeeper is
configured to run. It will be run in /etc. For example, "etckeeper vcs
diff" will run "git diff", etc.
.TP
.B uninit [-f]
This command DESTROYS DATA! It is the inverse of the init command, removing
VCS information and etckeeper's own bookkeeping information from the
directory. Use with caution. A typical use case would be to run etckeeper
uninit, then modify etckeeper.conf to use a different VCS, and then run
etckeeper init. (The -f switch can be used to force uninit without
prompting.)
.SH FILES
/usr/local/etc/etckeeper/etckeeper.conf is the configuration file.

/usr/local/etc/etckeeper also contains directories containing the programs that are
run for each of the above commands.
.SH ENVIRONMENT VARIABLES
ETCKEEPER_CONF_DIR path to configuration directory instead of default /usr/local/etc/etckeeper.
.SH SEE ALSO
/usr/share/doc/etckeeper/README.md.gz
.SH AUTHOR 
Joey Hess <id@joeyh.name>


CHANGELOG
etckeeper (1.18.21) upstream; urgency=medium

  * Consistently use mktemp if available, falling back to tempfile
    otherwise.
    Thanks, Adam Dinwoodie

 -- Joey Hess <id@joeyh.name>  Tue, 28 Nov 2023 14:09:16 -0400

etckeeper (1.18.20) upstream; urgency=medium

  * Fix a reversion in etckeeper init in version 1.18.19.
    Thanks, Georgy Yakovlev

 -- Joey Hess <id@joeyh.name>  Wed, 04 Jan 2023 13:23:19 -0400

etckeeper (1.18.19) upstream; urgency=medium

  * Added support for Gentoo (emerge, qlist, and cave)
    Thanks, Sam James
  * Skip running pre-commit hook inside linked worktrees,
    to avoid it updating .etckeeper with the permissions of files
    not in /etc.
    Thanks, Håkon Løvdal
  * commit: Run bzr with --quiet, since it outputs non-errors to stderr.
    Closes: #1018874

 -- Joey Hess <id@joeyh.name>  Mon, 02 Jan 2023 14:18:29 -0400

etckeeper (1.18.18) upstream; urgency=medium

  * Replace deprecated egrep with grep -E.
    Thanks, Sam James
  * Added support for Void Linux's xbps package manager.
    Thanks, Zev Weiss.

 -- Joey Hess <id@joeyh.name>  Thu, 08 Sep 2022 12:12:46 -0400

etckeeper (1.18.17) upstream; urgency=medium

  * Fix committing of files with spaces in name when perl is not available.
    Thanks, Henrik Riomar
  * Ignore udev's FHS violating large binary cache file /etc/udev/hwdb.bin
  * Avoid warning messages from grep about binary files when there are
    filenames in /etc that do not correspond to the current locale settings.
    Thanks, thm

 -- Joey Hess <id@joeyh.name>  Fri, 07 Jan 2022 11:15:07 -0400

etckeeper (1.18.16) upstream; urgency=medium

  * Improve sorting stability.
  * Prefer mktemp over tempfile as the latter displays a deprecation
    warning since debianutils 4.10.
    Thanks, Luke Mlsna.

 -- Joey Hess <id@joeyh.name>  Sat, 02 Jan 2021 11:31:23 -0400

etckeeper (1.18.15) upstream; urgency=medium

  * Use "command -v" rather than "which" to detect installed programs,
    as it is more portable.
    Thanks, Eli Schwartz.
  * Improve commit messages generated by package manager changes,
    listing packages that are responsible for the changed config files.
    Thanks to emkael for the patch.
  * If gc.auto is not configured, override the default to make it gc
    ten times more frequently, to avoid wasting space with loose objects.
  * update-ignore: Preserve permissions from any preexisting VCS ignore file.
    Thanks, Austin Chu.
  * Removed the debian directory from the upstream source package as it's
    not being maintained; see the debian package for an up-to-date one.
  * debian/changelog moved to CHANGELOG and debian/copyright to COPYRIGHT.

 -- Joey Hess <id@joeyh.name>  Mon, 23 Nov 2020 12:19:15 -0400

etckeeper (1.18.14) unstable; urgency=medium

  * pacman 5.2 deprecated File hooks, use Path.
    Thanks, Christian Hesse
  * Fix vcs subcommand setup for zsh completion.
    Thanks, James Rowe.

 -- Joey Hess <id@joeyh.name>  Wed, 22 Jan 2020 09:59:04 -0400

etckeeper (1.18.13) unstable; urgency=medium

  * Added zsh completion.
    Thanks, James Rowe
  * commit: Recent changes added code that does not work on all POSIX shells.
    Fixed by Thorsten Glaser.

 -- Joey Hess <id@joeyh.name>  Fri, 20 Dec 2019 16:19:09 -0400

etckeeper (1.18.12) unstable; urgency=medium

  * Fix bug in hostname determination in the previous release.
    Thanks, Christian Hesse

 -- Joey Hess <id@joeyh.name>  Fri, 29 Nov 2019 11:42:34 -0400

etckeeper (1.18.11) unstable; urgency=medium

  * Support platforms without a hostname command, fall back to 
    reading /etc/hostname.
    Thanks, Chris Morgan
  * commit: Support -mmessage, without a space, since eg git commit
    can be used that way.
    Thanks, martin f. krafft
  * commit: When multiple parameters are given, use them all as the commit
    message, instead of the old behavior of only using the first parameter and
    throwing the rest away.
    Thanks, martin f. krafft

 -- Joey Hess <id@joeyh.name>  Thu, 28 Nov 2019 18:14:46 -0400

etckeeper (1.18.10) unstable; urgency=medium

  * Avoid post-install failing when ps is from busybox or another
    version not supporting procps-specific options.
  * Use ps --no-headers rather than problimatic -h option.

 -- Joey Hess <id@joeyh.name>  Sun, 23 Dec 2018 13:05:44 -0400

etckeeper (1.18.9) unstable; urgency=medium

  * When run during a package installation, include in the commit
    message the command line that caused etckeeper to run.
    Thanks, Laszlo Gombos

 -- Joey Hess <id@joeyh.name>  Wed, 12 Dec 2018 01:01:05 -0400

etckeeper (1.18.8) unstable; urgency=medium

  * Work around git commit's lack of robustness, by providing
    reasonable default values for GIT_COMMITTER_EMAIL etc.
    This was already done as part of the su/sudo handling,
    and is now always done.
  * Don't hardcode the master branch when pushing to PUSH_REMOTE.
    Instead, let git push whatever branches it is configured to
    push to that remote.

 -- Joey Hess <id@joeyh.name>  Tue, 05 Jun 2018 16:00:41 -0400

etckeeper (1.18.7) unstable; urgency=medium

  * Added some unit tests.
    Thanks, Henrik Riomar.
  * etckeeper will work on systems that do not have perl installed.
    (perl is still used when available as it's faster)
    Thanks, William Johansson and radhus.
  * Prevent LC_ALL overriding the LC_COLLATE used to sort metadata.

 -- Joey Hess <id@joeyh.name>  Thu, 08 Jun 2017 12:58:20 -0400

etckeeper (1.18.6) unstable; urgency=medium

  * Only show errors (no progress indicators) when pushing Git/Mercurial repos
    to avoid unncessary cron mails.
    Thanks, Nils Steinger.
  * Fix regex in 20-warn-problem-files.
  * Added support for apk (alpine linux)
    Thanks, Henrik Riomar.

 -- Joey Hess <id@joeyh.name>  Sun, 29 Jan 2017 14:37:24 -0400

etckeeper (1.18.5) unstable; urgency=medium

  * Make etckeeper commit store metadata changes. The pre-commit
    hook has always (and continues) to do that, but pre-commit is only

pkgng/etckeeper.c
/*
 * Copyright (c) 2015 William Johansson <radar@radhuset.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer
 *    in this position and unchanged.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * Based upon the pkg plugin zfssnap:
 * Copyright (c) 2012 Marin Atanasov Nikolov <dnaeon@gmail.com>
 */

#include <stdio.h>
#include <sys/wait.h>
#include <err.h>
#include <errno.h>
#include <spawn.h>

#include <pkg.h>
#define PLUGIN_NAME "etckeeper"
/* TODO: make configuration param? */
/* A wrapper script is used for multiple target dirs. */
#define ETCKEEPER_PATH "/usr/local/libexec/etckeeper/etckeeper-dirs"

extern char **environ;

struct pkg_plugin *self;

static int pre_install_hook(void *data, struct pkgdb *db);
static int post_install_hook(void *data, struct pkgdb *db);

int
pkg_plugin_init(struct pkg_plugin *p)
{
	self = p;

	pkg_plugin_set(p, PKG_PLUGIN_NAME, PLUGIN_NAME);
	pkg_plugin_set(p, PKG_PLUGIN_DESC, "etckeeper plugin");
	pkg_plugin_set(p, PKG_PLUGIN_VERSION, "1.0.0");

	/* NOTE: upgrade/deinstall is regarded as install */

	if (pkg_plugin_hook_register(p, PKG_PLUGIN_HOOK_PRE_INSTALL, &pre_install_hook) != EPKG_OK) {
		pkg_plugin_error(self, "failed to hook into the library");
		return (EPKG_FATAL);
	}

	if (pkg_plugin_hook_register(p, PKG_PLUGIN_HOOK_POST_INSTALL, &post_install_hook) != EPKG_OK) {
		pkg_plugin_error(self, "failed to hook into the library");
		return (EPKG_FATAL);
	}

	if (pkg_plugin_hook_register(p, PKG_PLUGIN_HOOK_PRE_DEINSTALL, &pre_install_hook) != EPKG_OK) {
		pkg_plugin_error(self, "failed to hook into the library");
		return (EPKG_FATAL);
	}

	if (pkg_plugin_hook_register(p, PKG_PLUGIN_HOOK_POST_DEINSTALL, &post_install_hook) != EPKG_OK) {
		pkg_plugin_error(self, "failed to hook into the library");
		return (EPKG_FATAL);
	}

	if (pkg_plugin_hook_register(p, PKG_PLUGIN_HOOK_PRE_UPGRADE, &pre_install_hook) != EPKG_OK) {
		pkg_plugin_error(self, "failed to hook into the library");
		return (EPKG_FATAL);
	}

	if (pkg_plugin_hook_register(p, PKG_PLUGIN_HOOK_POST_UPGRADE, &post_install_hook) != EPKG_OK) {
		pkg_plugin_error(self, "failed to hook into the library");
		return (EPKG_FATAL);
	}

	return (EPKG_OK);
}

static int
call_etckeeper(char *arg)
{
	int error, pstat;
	pid_t pid;
	char *argv[] = {
		"etckeeper",
		arg,
		NULL,
	};

	if ((error = posix_spawn(&pid, ETCKEEPER_PATH, NULL, NULL,
		__DECONST(char **, argv), environ)) != 0) {
		errno = error;
		pkg_plugin_errno(self, "Failed to spawn process", arg);
		return (EPKG_FATAL);
	}
	while (waitpid(pid, &pstat, 0) == -1) {
		if (errno != EINTR)
			return (EPKG_FATAL);
	}

	if ((error = WEXITSTATUS(pstat)) != 0) {
		pkg_plugin_error(self, "etckeeper failed with exit code %d", error);
		return (EPKG_FATAL);
	}

	return (EPKG_OK);
}

static int
pre_install_hook(void *data, struct pkgdb *db)
{
	return call_etckeeper("pre-install");
}

static int
post_install_hook(void *data, struct pkgdb *db)
{
	return call_etckeeper("post-install");
}
